### PR TITLE
Perl wrapper of c bivarivate cdf function

### PR DESCRIPTION
**Pull request by shahrizal-binary**

### CODE REVIEW

File: lib/Math/BivariateCDF.pm [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+Math::BivariateCDF - Perl extension for Bivariate CDF functions.
+
+=head1 SYNOPSIS
+
+  use Math::BivariateCDF;
+
+=head1 DESCRIPTION
+
+Stub documentation for Math::BivariateCDF, created by h2xs.
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: please, update

File: typemap [hunk 2]
```diff
@@ -0,0 +1,2 @@
+double *					T_PTROBJ
```
Comment (basiliscos) [inline typemap]: IMHO you should not export the double\* at all, but use native perl type for floats `NV`

File: lib/Math/BivariateCDF.pm [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+#XXX	if ($] >= 5.00561) {
+#XXX	    *$AUTOLOAD = sub () { $val };
+#XXX	}
+#XXX	else {
+	    *$AUTOLOAD = sub { $val };
+#XXX	}
+    }
+    goto &$AUTOLOAD;
+}
+
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm:30]: Please, don't use autoload. It's crap in Perl.

File: Makefile.PL [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
+    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
+      (ABSTRACT_FROM  => 'lib/Math/BivariateCDF.pm', # retrieve abstract from module
+       AUTHOR         => 'root <root@>') : ()),
+    LIBS              => ['-lm'], # e.g., '-lm'
+    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
+    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
+    OBJECT            => '$(O_FILES)', # link all the C files too
+);
+if  (eval {require ExtUtils::Constant; 1}) {
```
Comment (basiliscos) [inline Makefile.PL]: I didn't understand at all - why do u need constants at all? If they are not exposed to perl at all, then, keep all them in C-code / C-headers.

File: Makefile.PL [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+use ExtUtils::MakeMaker;
+# See lib/ExtUtils/MakeMaker.pm for details of how to influence
+# the contents of the Makefile that is written.
+WriteMakefile(
+    NAME              => 'Math::BivariateCDF',
+    VERSION_FROM      => 'lib/Math/BivariateCDF.pm', # finds $VERSION
+    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
+    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
+      (ABSTRACT_FROM  => 'lib/Math/BivariateCDF.pm', # retrieve abstract from module
+       AUTHOR         => 'root <root@>') : ()),
```
Comment (basiliscos) [inline Makefile.PL]: Omg :) Please check the code and clean up from unneeded comments

File: lib/Math/BivariateCDF.pm [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+=head1 SYNOPSIS
+
+  use Math::BivariateCDF;
+
+  Math::BivariateCDF::bivnor( $x, $y, $z );
+
+=head1 DESCRIPTION
+
+This is a Perl wrapper for TOMS462 C library which evaluates the upper right tail of the bivariate normal distribution; that is, the probability that normal variables X and Y with correlation R will satisfy H <= X and K <= Y.
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: Please, don't use long lines

File: lib/Math/BivariateCDF.pm [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  void bivariate_normal_cdf_values ( int *n_data, double *x, double *y,
+  double *r, double *fxy )
+  double bivnor ( double ah, double ak, double r )
+  double gauss ( double t )
+  double r8_abs ( double x )
+  double r8_max ( double x, double y )
+  double r8_min ( double x, double y )
+  void timestamp ( void )
+
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm:62]: ??? 

`void timestamp(void)` What is is?

Please, remove all unexported functions from here. I think we nee only one function - `bivariate_normal_cdf_values`
all other should be removed

File: lib/Math/BivariateCDF.pm [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+https://people.sc.fsu.edu/~jburkardt/c_src/toms462/toms462.html
+
+=head1 AUTHOR
+
+binary.com, E<lt>support@binary.com<gt>
+
+=head1 COPYRIGHT AND LICENSE
+
+Copyright (C) 2016 by root
+
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm:76]: :)

File: const-c.inc [hunk 9]
```diff
@@ -0,0 +1,55 @@
+#define PERL_constant_NOTFOUND	1
```
Comment (basiliscos) [inline const-c.inc]: I don't know what this file does. Remove it, please of explain the need of it.

File: const-xs.inc [hunk 10]
```diff
@@ -0,0 +1,89 @@
+void
+constant(sv)
+    PREINIT:
+#ifdef dXSTARG
+	dXSTARG; /* Faster if we have it.  */
+#else
+	dTARGET;
```
Comment (basiliscos) [inline const-xs.inc]: I don't know what this file does. Remove it, please of explain the need of it.

File: fallback/const-c.inc [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#define PERL_constant_NOTFOUND	1
```
Comment (basiliscos) [inline fallback/const-c.inc]: I don't know what this file does. Remove it, please of explain the need of it.

File: fallback/const-xs.inc [hunk 12]
```diff
@@ -0,0 +1,89 @@
+void
```
Comment (basiliscos) [inline fallback/const-xs.inc]: I don't know what this file does. Remove it, please of explain the need of it.

File: lib/Math/BivariateCDF.pm [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+# Items to export into callers namespace by default. Note: do not export
+# names by default without a very good reason. Use EXPORT_OK instead.
+# Do not simply export all your public functions/methods/constants.
+
+# This allows declaration	use Math::BivariateCDF ':all';
+# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
+# will save memory.
+our %EXPORT_TAGS = ( 'all' => [ qw(
+	bivariate_normal_cdf_values
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: It is better to name the function `bivariate_normal_cdf`

File: t/Math-BivariateCDF.t [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# Before `make install' is performed this script should be runnable with
+# `make test'. After `make install' it should work as `perl Math-BivariateCDF.t'
+
+#########################
+
+# change 'tests => 1' to 'tests => last_test_to_print';
+
```
Comment (basiliscos) [inline t/Math-BivariateCDF.t]: Please, remove unneeded comments.

File: lib/Math/BivariateCDF.pm [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+binary.com, E<lt>support@binary.com<gt>
+
+=head1 COPYRIGHT AND LICENSE
+
+Copyright (C) 2016 by root
+
+This library is free software; you can redistribute it and/or modify
+it under the same terms as Perl itself, either Perl version 5.14.2 or,
+at your option, any later version of Perl 5 you may have available.
+
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm:78]: You violate licence of `toms462`.

`This code is distributed under the GNU LGPL license.`

The module should be distributed under LGPL too.

File: toms462.c [hunk 16]
```diff
@@ -0,0 +1,634 @@
+# include <stdlib.h>
+# include <stdio.h>
+# include <math.h>
+# include <time.h>
+
+# include "toms462.h"
+
+/******************************************************************************/
+
+void bivariate_normal_cdf_values ( int *n_data, double *x, double *y,
```
Comment (basiliscos) [inline toms462.c:10]: I think you need to create an XS-wrapper around that function

File: lib/Math/BivariateCDF.pm (hunk 7 above)
Comment (shahrizal-binary) [inline lib/Math/BivariateCDF.pm:62]: We only use bivnor . I will remove the rest.

File: const-c.inc (hunk 9 above)
Comment (shahrizal-binary) [inline const-c.inc]: The .inc files are auto generated. I will try remove them and see what happen.

File: lib/Math/BivariateCDF.pm (hunk 13 above)
Comment (shahrizal-binary) [inline lib/Math/BivariateCDF.pm]: We only use the function called bivnor . I will remove the rest.

File: const-c.inc (hunk 9 above)
Comment (basiliscos) [inline const-c.inc]: You are responsible for every line of code, because the PR states, that Shahrizal is an author :) That's why it is better to avoid all autogenerators. In any case, I'll will not accept the PR, if you'll not be able to explain the need of any line of code.

Comment (shahrizal-binary) [inline const-c.inc]: Will play around with these files and see what it does.

File: toms462.c (hunk 16 above)
Comment (shahrizal-binary) [inline toms462.c:10]: Not this one, the one we are using is called 'bivnor' .

File: Makefile.PL [hunk 17]
```diff
@@ -1,23 +1,17 @@
 use 5.014002;
```
Comment (basiliscos) [inline Makefile.PL]: that isn't need too :)

File: lib/Math/BivariateCDF.pm [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+=head1 NAME
+
+Math::BivariateCDF - Perl extension for Bivariate CDF functions.
+
+=head1 SYNOPSIS
+
+  use Math::BivariateCDF;
+
+  Math::BivariateCDF::bivnor( $x, $y, $z );
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: Oh, that's shit!

Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: It should be like:

``` perl
use Math::BivariateCDF qw/bivnor/;

my $variation = bivnor(1, 2, 3);

print("variation = $variation\n"); # should be 0.5

```

I.e. provide a numbers (input arguments) + comment their meaning (or reference to their meaning);  Please, substitute the numbers 1,2,3 with something meaningful for for the function.

**Comment by basiliscos** [comment]
Looks very good!

**Comment by basiliscos** [comment]
Please, add tests, including tests for edge cases, and for invalid input (e.g. -1 or string); because . c-functions like to crash on invalid input, but to crash perl-interpreter is unacceptable, i.e .invalid values should be checked somewhere and error returned.

File: Makefile.PL [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    VERSION_FROM      => 'lib/Math/BivariateCDF.pm',
+    CONFIGURE_REQUIRES => {
+                "ExtUtils::MakeMaker" => 0,
+    },
+    BUILD_REQUIRES => {
+                'Test::More'      => 0,
+                'Test::Warnings'  => 0,
+    },
+    PREREQ_PM => {
+                "ExtUtils::MakeMaker" => 0,
```
Comment (basiliscos) [inline Makefile.PL]: Remove this one, please

File: Makefile.PL [hunk 20]
```diff
@@ -0,0 +1,27 @@
+use ExtUtils::MakeMaker;
+
+
+WriteMakefile(
+    NAME              => 'Math::BivariateCDF',
+    VERSION_FROM      => 'lib/Math/BivariateCDF.pm',
+    CONFIGURE_REQUIRES => {
+                "ExtUtils::MakeMaker" => 0,
+    },
+    BUILD_REQUIRES => {
```
Comment (basiliscos) [inline Makefile.PL]: this should be `TEST_REQUIRES`

File: lib/Math/BivariateCDF.pm [hunk 21]
```diff
@@ -0,0 +1,83 @@
+package Math::BivariateCDF;
+
+use 5.014002;
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: please, remove this line

File: lib/Math/BivariateCDF.pm [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+use strict;
+use warnings;
+use Carp;
+
+require Exporter;
+use AutoLoader;
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm]: Please, remove AutoLoader.

File: lib/Math/BivariateCDF.pm [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+  use Math::BivariateCDF qw(bivnor);
+
+  my $cdf = Math::BivariateCDF::bivnor( 0.036412293026205, 0.0708220179920969, 0.734208757779421 );
+
+  print $cdf; #0.359905409368831
+
+=head1 DESCRIPTION
+
+This is a Perl wrapper for TOMS462 C library which evaluates the upper right tail of the bivariate normal distribution.
```
Comment (basiliscos) [inline lib/Math/BivariateCDF.pm:53]: Some link to wikipedia, explaining, what is  Bivariate CDF would be nice.

File: toms462.c [hunk 24]
```diff
@@ -0,0 +1,634 @@
...
+  }
+
+  rr = ( 1.0 + r ) * ( 1.0 - r );
+
+  if ( rr < 0.0 )
+  {
+    fprintf ( stderr, "\n" );
+    fprintf ( stderr, "BIVNOR - Fatal error!\n" );
+    fprintf ( stderr, "  1 < |R|.\n" );
+    exit ( 0 );
```
Comment (basiliscos) [inline toms462.c:261]: This is absolutely unacceptable in XS-code.

Please, check the pre-conditions in XS code wrapper, to be sure, that they'll never happen (rr < 0), and if they do, then just do not call that function, but die with error.

When you do preconditions check, please, comment (but not remove) this lines, that they are checked.

File: lib/Math/BivariateCDF.pm (hunk 23 above)
Comment (shahrizal-binary) [inline lib/Math/BivariateCDF.pm:53]: Added link.

File: toms462.c (hunk 24 above)
Comment (shahrizal-binary) [inline toms462.c:261]: Added check in XS.
